<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Description &mdash; mpieva/quicksand v1.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/style.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Quickstart" href="quickstart.html" />
    <link rel="prev" title="Quicksand&#39;s documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> mpieva/quicksand
          </a>
              <div class="version">
                v1.6
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Description</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#workflow">Workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#estimatecrosscontamination">EstimateCrossContamination</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastq2bam">fastq2bam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiterbam">fiterBam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filterlength">filterLength</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tofasta">toFasta</a></li>
<li class="toctree-l3"><a class="reference internal" href="#runkrakenuniq">runKrakenUniq</a></li>
<li class="toctree-l3"><a class="reference internal" href="#findbestnode">findBestNode</a></li>
<li class="toctree-l3"><a class="reference internal" href="#extractbam">extractBam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mapbwa">mapBWA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filtermappedbam">filterMappedBam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dedupbam">dedupBam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#runintersectbed">runIntersectBed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#analyzedeamination">analyzeDeamination</a></li>
<li class="toctree-l3"><a class="reference internal" href="#extractdeaminatedreads">extractDeaminatedReads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#maskdeamination">maskDeamination</a></li>
<li class="toctree-l3"><a class="reference internal" href="#creatempileups">createMpileups</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="quicksand-build.html">quicksand-build</a></li>
<li class="toctree-l1"><a class="reference internal" href="contribute.html">Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">mpieva/quicksand</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Description</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/description.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="description">
<h1>Description<a class="headerlink" href="#description" title="Permalink to this headline"></a></h1>
<a class="reference internal image-reference" href="_images/pipeline_overview_v1.5.png"><img alt="Graphical overview over the processes of the quicksand pipeline" class="align-center" src="_images/pipeline_overview_v1.5.png" style="width: 800px;" /></a>
<p>This section describes the processes implemented in the <strong>quicksand</strong> pipeline. See <a class="reference internal" href="quicksand-build.html#quicksand-build-page"><span class="std std-ref">quicksand-build</span></a> for a description of
the quicksand-build pipeline</p>
<section id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline"></a></h2>
<p>The basic workflow of the quicksand pipline consists of a <strong>metagenomic classification</strong>, a <strong>mapping</strong> and a basic <strong>analysis</strong> of the alignment. The pipeline
produces taxonomic profiles for the analyzed input files, as well as the binned <code class="file docutils literal notranslate"><span class="pre">bam</span></code> files, grouped by redgroup and taxonomy for all levels of the
pipeline - as shown in the figure above. Each process is in detail described in the following chapters</p>
<section id="estimatecrosscontamination">
<h3>EstimateCrossContamination<a class="headerlink" href="#estimatecrosscontamination" title="Permalink to this headline"></a></h3>
<p>WIP</p>
</section>
<section id="fastq2bam">
<h3>fastq2bam<a class="headerlink" href="#fastq2bam" title="Permalink to this headline"></a></h3>
<p>All <code class="file docutils literal notranslate"><span class="pre">fastq</span></code> files entering the pipeline are converted to unpaired and unmapped <code class="file docutils literal notranslate"><span class="pre">bam</span></code> files using <code class="code docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">import</span> <span class="pre">-0</span></code>.
In case of paired end <code class="file docutils literal notranslate"><span class="pre">fastq</span></code> files, make sure to merge them before using the pipeline.</p>
</section>
<section id="fiterbam">
<h3>fiterBam<a class="headerlink" href="#fiterbam" title="Permalink to this headline"></a></h3>
<p><code class="file docutils literal notranslate"><span class="pre">Bam</span></code> files are filtered based on the provided filterflag using <code class="code docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">view</span> <span class="pre">-F</span></code>. By default paired-end reads are removed from the
<code class="file docutils literal notranslate"><span class="pre">bam</span></code> file.</p>
</section>
<section id="filterlength">
<h3>filterLength<a class="headerlink" href="#filterlength" title="Permalink to this headline"></a></h3>
<p>Using a custom <code class="code docutils literal notranslate"><span class="pre">bam-lengthfilter</span></code> package, sequences are removed from the <code class="file docutils literal notranslate"><span class="pre">bam</span></code> file that fall below the <code class="code docutils literal notranslate"><span class="pre">--bamfilter_length_cutoff</span></code> threshold (default: 35bp)</p>
</section>
<section id="tofasta">
<h3>toFasta<a class="headerlink" href="#tofasta" title="Permalink to this headline"></a></h3>
<p>As preparation for the krakenuniq run, <code class="file docutils literal notranslate"><span class="pre">bam</span></code> files are converted to <code class="file docutils literal notranslate"><span class="pre">fasta</span></code> files</p>
</section>
<section id="runkrakenuniq">
<h3>runKrakenUniq<a class="headerlink" href="#runkrakenuniq" title="Permalink to this headline"></a></h3>
<p>Sequences contained in the <code class="file docutils literal notranslate"><span class="pre">fasta</span></code> file are classified by the metagenomic classifier <code class="code docutils literal notranslate"><span class="pre">krakenuniq</span></code>. Quicksand uses krakenuniq with a precompiled
database created from the current non-redundant mtDNA RefSeq database with a default kmer size of 22 (see <a class="reference internal" href="quicksand-build.html#quicksand-build-page"><span class="std std-ref">quicksand-build</span></a> or <a class="reference internal" href="installation.html#setup"><span class="std std-ref">Create datastructure</span></a>).
The speed of krakenuniq allows for a quick sorting of sequences into families. To filter out false-positive assignments, families are
removed from the assignment falling below the minimum number of reads (<code class="code docutils literal notranslate"><span class="pre">--krakenuniq_min_reads</span></code>) and the minimal number of kmers (<code class="code docutils literal notranslate"><span class="pre">--krakenuniq_min_kmers</span></code>)
on the family node. The result of this process is a taxonomic profile of the readgroup.</p>
</section>
<section id="findbestnode">
<h3>findBestNode<a class="headerlink" href="#findbestnode" title="Permalink to this headline"></a></h3>
<p>This process parses the kraken-reports. For each assigned family reported by krakenuniq, the node with the highest number of assigned unique kmers is
picked as the taxon representative for that family.</p>
</section>
<section id="extractbam">
<h3>extractBam<a class="headerlink" href="#extractbam" title="Permalink to this headline"></a></h3>
<p>Using the kraken-report and the length-filtered <code class="file docutils literal notranslate"><span class="pre">bam</span></code> file, this process collects all sequences assigned to one clade into a new <code class="file docutils literal notranslate"><span class="pre">bam</span></code>
file. Extraction happens either on the family or order-level, as specified with the <code class="code docutils literal notranslate"><span class="pre">--taxlvl</span></code> flag, using the custom <code class="code docutils literal notranslate"><span class="pre">bamfilter</span></code> package.</p>
</section>
<section id="mapbwa">
<h3>mapBWA<a class="headerlink" href="#mapbwa" title="Permalink to this headline"></a></h3>
<p>The extracted sequences are mapped against all the reference genomes of species belonging to the 'bestNode' found in the 'findBestNode' process
using the <code class="code docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">bam2bam</span></code> command of the <a class="reference external" href="https://github.com/mpieva/network-aware-bwa">network-aware fork</a> of BWA with
ancient parameters (<code class="code docutils literal notranslate"><span class="pre">n</span> <span class="pre">0.01</span> <span class="pre">-o</span> <span class="pre">2</span> <span class="pre">-l</span> <span class="pre">16500</span></code>). Unmapped sequences or sequences with a mapping quality of less than 25 are removed from the alignment</p>
</section>
<section id="filtermappedbam">
<h3>filterMappedBam<a class="headerlink" href="#filtermappedbam" title="Permalink to this headline"></a></h3>
<p>Mapped bam files are filtered for the set alignment quality score</p>
</section>
<section id="dedupbam">
<h3>dedupBam<a class="headerlink" href="#dedupbam" title="Permalink to this headline"></a></h3>
<p>Exact PCR-duplicates are collapsed into unique sequences using <a class="reference external" href="https://github.com/mpieva/biohazard-tools">bam-rmdup</a> based on the sharing of identical
alignment start and end coordinates. From all mapped genomes, the one with the highest numbers of basepairs covered is picked as _the_
representative species for the subsequent steps.</p>
</section>
<section id="runintersectbed">
<h3>runIntersectBed<a class="headerlink" href="#runintersectbed" title="Permalink to this headline"></a></h3>
<p>The deduped alignments are then depleted of reads that overlap sites marked as non-informative by <code class="code docutils literal notranslate"><span class="pre">dustmasker</span></code>. That step is skipped
for families with a fixed reference genome (see <code class="code docutils literal notranslate"><span class="pre">--fixed</span></code> flag)</p>
</section>
<section id="analyzedeamination">
<h3>analyzeDeamination<a class="headerlink" href="#analyzedeamination" title="Permalink to this headline"></a></h3>
<p>The final step(s) in the pipeline look for C to T substitutions in the query sequences in respect to the aligned reference genome.
Ancient DNA shows characteristic C to T substitutions at the 3’ and 5’ ends
of DNA fragments - a degradation pattern used to identify ancient DNA. Families which sequences show more than 10% of terminal C bases in the
reference genome replaced by a T are reported as being ancient (++).</p>
</section>
<section id="extractdeaminatedreads">
<h3>extractDeaminatedReads<a class="headerlink" href="#extractdeaminatedreads" title="Permalink to this headline"></a></h3>
<p>For families with a fixed reference genome (see <code class="code docutils literal notranslate"><span class="pre">--fixed</span></code> flag), extract deaminated reads into two different <code class="code docutils literal notranslate"><span class="pre">.bam</span></code> files. One file with sequences that
show C to T substitutions on one of the terminal base pairs and one file that looks at the first and last 3 terminal base positions.</p>
</section>
<section id="maskdeamination">
<h3>maskDeamination<a class="headerlink" href="#maskdeamination" title="Permalink to this headline"></a></h3>
<p>For families with a fixed reference genome (see <code class="code docutils literal notranslate"><span class="pre">--fixed</span></code> flag), take the extracted deaminated reads and set the alignment quality score of the last
three terminal T bases to 15.</p>
</section>
<section id="creatempileups">
<h3>createMpileups<a class="headerlink" href="#creatempileups" title="Permalink to this headline"></a></h3>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Quicksand&#39;s documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="quickstart.html" class="btn btn-neutral float-right" title="Quickstart" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Merlin Szymanski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>