<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Installation &mdash; mpieva/quicksand v1.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/style.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Usage" href="usage.html" />
    <link rel="prev" title="Quickstart" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> mpieva/quicksand
          </a>
              <div class="version">
                v1.6
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="description.html">Description</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#test-runability">Test runability</a></li>
<li class="toctree-l2"><a class="reference internal" href="#container-cache">Container Cache</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-datastructure">Create datastructure</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-pipeline">Run the pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filter-the-results">Filter the Results</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="quicksand-build.html">quicksand-build</a></li>
<li class="toctree-l1"><a class="reference internal" href="contribute.html">Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">mpieva/quicksand</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Installation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/installation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="installation">
<span id="install-page"></span><h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline"></a></h1>
<p>Use this section to set up the pipeline to work with real data. To test if the pipeline works on your computer, see the <a class="reference internal" href="quickstart.html#quickstart-page"><span class="std std-ref">Quickstart</span></a> section</p>
<section id="requirements">
<span id="id1"></span><h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline"></a></h2>
<p>Make sure you have the following two components installed.</p>
<dl class="field-list simple">
<dt class="field-odd">Nextflow</dt>
<dd class="field-odd"><p>See more details here: <a class="reference external" href="https://www.nextflow.io/docs/latest/getstarted.html">Nextflow</a></p>
</dd>
<dt class="field-even">Singularity or Docker</dt>
<dd class="field-even"><p>See more details here: <a class="reference external" href="https://sylabs.io/guides/3.0/user-guide/installation.html">Singularity</a> or <a class="reference external" href="https://docs.docker.com/get-docker/">Docker</a></p>
</dd>
</dl>
<p>Nextflow is the pipeline framework, while Singularity/Docker are software tools for the containerization of code.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>check the successful installation of the software by running:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="o">-</span><span class="n">v</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">nextflow</span> <span class="n">version</span> <span class="mf">20.04.1.5335</span>
<span class="n">singularity</span> <span class="o">--</span><span class="n">version</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">singularity</span> <span class="n">version</span> <span class="mf">3.7.2</span><span class="o">-</span><span class="n">dirty</span>
</pre></div>
</div>
</div>
<p><span class="heading1">Working with test data</span></p>
</section>
<section id="test-runability">
<h2>Test runability<a class="headerlink" href="#test-runability" title="Permalink to this headline"></a></h2>
<p>To test the runability of the pipeline, see the <a class="reference internal" href="quickstart.html#quickstart-page"><span class="std std-ref">Quickstart</span></a> section</p>
</section>
<section id="container-cache">
<span id="container"></span><h2>Container Cache<a class="headerlink" href="#container-cache" title="Permalink to this headline"></a></h2>
<p>Quicksand uses container software (Docker,Singularity) to ensure the pipeline runs within a stable environment.
For single-tool processes, container images are either pulled from the <a class="reference external" href="https://depot.galaxyproject.org/singularity">Galaxy image repository</a> or the
<a class="reference external" href="https://quay.io/organization/biocontainers">Quai.io biocontainer repository</a>. For multi-tool processes and
custom functions, self-built images are hosted on <a class="reference external" href="https://hub.docker.com/r/merszym">Dockerhub</a>. The
underlying Dockerfiles for those images can be found within the <code class="file docutils literal notranslate"><span class="pre">assets/docker</span></code> directory of the <a class="reference external" href="https://www.github.com/mpieva/quicksand">repository</a>.</p>
<p>To reuse downloaded images for multiple pipeline runs, specify a cache-directory. type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>mkdir singularity
export NXF_SINGULARITY_CACHEDIR=$PWD/singularity
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Make sure that Singularity has the permissions to overlay and access your file-system within the
container. Otherwise the pipeline wont be able to read from or write to it. Upon a
<span class="bold">&quot;file doesn't exist&quot; error</span>, create an additional <code class="file docutils literal notranslate"><span class="pre">nextflow.config</span></code> config-file</p>
<p>Add the following content to the file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">singularity</span> <span class="p">{</span>
  <span class="n">runOptions</span> <span class="o">=</span> <span class="s2">&quot;--bind /directory/in/use&quot;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Add this file to your nextflow run with the <code class="code docutils literal notranslate"><span class="pre">-c</span></code> flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="o">...</span> <span class="o">-</span><span class="n">c</span> <span class="n">singularity</span><span class="o">/</span><span class="n">nextflow</span><span class="o">.</span><span class="n">config</span>
</pre></div>
</div>
</div>
<div class="admonition attention">
<p class="admonition-title">Attention</p>
<p>the <code class="code docutils literal notranslate"><span class="pre">-profile</span></code> and the <code class="code docutils literal notranslate"><span class="pre">-c</span></code> flag have only one dash!</p>
</div>
<p>Now -again- the pipeline can be tested by running:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="o">-</span><span class="n">profile</span> <span class="n">test</span><span class="p">,</span><span class="n">singularity</span> <span class="o">-</span><span class="n">c</span> <span class="n">singularity</span><span class="o">/</span><span class="n">nextflow</span><span class="o">.</span><span class="n">config</span>
</pre></div>
</div>
<div class="line-block">
<div class="line">The meaning of the flags and the different ways of customizing the pipeline is described in the <a class="reference internal" href="usage.html#usage-page"><span class="std std-ref">Usage</span></a> section.</div>
<div class="line">In case of choosing Docker over Singularity, use the <code class="code docutils literal notranslate"><span class="pre">-profile</span> <span class="pre">test,docker</span></code> command.</div>
</div>
<p>The <code class="file docutils literal notranslate"><span class="pre">singularity</span></code> directory should now contain all the images used in the pipeline:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>singularity
├── depot.galaxyproject.org-singularity-samtools-1.15.1--h1170115_0.img
├── merszym-biohazard_bamrmdup-v0.2.img
└── merszym-quicksand-1.2.img
</pre></div>
</div>
<p><span class="heading1">Working with real data</span></p>
</section>
<section id="create-datastructure">
<span id="setup"></span><h2>Create datastructure<a class="headerlink" href="#create-datastructure" title="Permalink to this headline"></a></h2>
<p>The required underlying datastructure of the pipeline is in detail described in the <a class="reference internal" href="quicksand-build.html#quicksand-build-page"><span class="std std-ref">quicksand-build</span></a> section</p>
<p>In short: You need a <span class="bold">precompiled kraken database</span>, the respective <span class="bold">reference genomes</span> and <span class="bold">bedfiles</span> indicating low-complexity regions.
Use the supplementary pipeline <code class="code docutils literal notranslate"><span class="pre">quicksand-build</span></code> (once) to download the taxonomy from NCBI/taxonomy, all mitochondrial
genomes from NCBI/RefSeq and create the required databases and files for you.</p>
<p>For this session create the datastructure for the <span class="bold">Primate mtDNA</span> from RefSeq:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">mpieva</span><span class="o">/</span><span class="n">quicksand</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">outdir</span> <span class="n">refseq</span> <span class="o">--</span><span class="n">include</span> <span class="n">Primates</span>
</pre></div>
</div>
<div class="admonition attention">
<p class="admonition-title">Attention</p>
<div class="line-block">
<div class="line">Building the database requires ~40G of RAM</div>
<div class="line">Be patient, downloading the taxonomy plus the creation of the database might take <span class="bold">~1h</span>.</div>
</div>
</div>
<p>This command creates a directory <code class="file docutils literal notranslate"><span class="pre">refseq</span></code> that contains the files required to run quicksand:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>refseq
├── kraken
│    └── Mito_db_kmer22
├── genomes
│    ├── {family}
│    │    └── {species}.fasta
│    └── taxid_map.tsv
└── masked
     └── {species}_masked.bed
</pre></div>
</div>
<p>With the datastructure created, the pipeline is ready to be used with the following flags:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">--</span><span class="n">db</span>         <span class="n">refseq</span><span class="o">/</span><span class="n">kraken</span><span class="o">/</span><span class="n">Mito_db_kmer22</span><span class="o">/</span>
<span class="o">--</span><span class="n">genomes</span>    <span class="n">refseq</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span>
<span class="o">--</span><span class="n">bedfiles</span>   <span class="n">refseq</span><span class="o">/</span><span class="n">masked</span><span class="o">/</span>
</pre></div>
</div>
</section>
<section id="run-the-pipeline">
<h2>Run the pipeline<a class="headerlink" href="#run-the-pipeline" title="Permalink to this headline"></a></h2>
<p>As <span class="bold">input</span> for the pipeline, download the Hominin &quot;Hohlenstein-Stadel&quot; mtDNA <a class="footnote-reference brackets" href="#id3" id="id2">1</a> into a directory <span class="bold">split</span></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="o">-</span><span class="n">P</span> <span class="n">split</span> <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">eva</span><span class="o">.</span><span class="n">mpg</span><span class="o">.</span><span class="n">de</span><span class="o">/</span><span class="n">neandertal</span><span class="o">/</span><span class="n">Hohlenstein</span><span class="o">-</span><span class="n">Stadel</span><span class="o">/</span><span class="n">BAM</span><span class="o">/</span><span class="n">mtDNA</span><span class="o">/</span><span class="n">HST</span><span class="o">.</span><span class="n">raw_data</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>And run the quicksand pipeline:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">mpieva</span><span class="o">/</span><span class="n">quicksand</span> \
    <span class="o">--</span><span class="n">db</span>        <span class="n">refseq</span><span class="o">/</span><span class="n">kraken</span><span class="o">/</span><span class="n">Mito_db_kmer22</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">genomes</span>   <span class="n">refseq</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">bedfiles</span>  <span class="n">refseq</span><span class="o">/</span><span class="n">masked</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">split</span>     <span class="n">split</span> \
    <span class="o">-</span><span class="n">profile</span>    <span class="n">singularity</span>
</pre></div>
</div>
<div class="line-block">
<div class="line">Please see the <a class="reference internal" href="usage.html#usage-page"><span class="std std-ref">Usage</span></a> section for an explaination of the flags and the input!</div>
<div class="line">Please see the <a class="reference internal" href="usage.html#output"><span class="std std-ref">Output</span></a> section for an explaination of the output files!</div>
</div>
<p>A summary of all the stats can be found in the <code class="file docutils literal notranslate"><span class="pre">final_report.tsv</span></code> file</p>
</section>
<section id="filter-the-results">
<h2>Filter the Results<a class="headerlink" href="#filter-the-results" title="Permalink to this headline"></a></h2>
<p>As can be seen in the <code class="code docutils literal notranslate"><span class="pre">final_report.tsv</span></code>, not all sequences were assigned to Homindae, but to a couple of other Primate families too.
The assignment of false positive taxa is a well-known problem of kmer-based assignment methods and additional filters need to be applied.</p>
<p>Based on simulated data, our recommended cutoffs are:</p>
<ul class="simple">
<li><p><span class="bold">FamPercentage</span> cutoff of 1% and/or</p></li>
<li><p><span class="bold">ProportionMapped</span> cutoff of 0.5-0.7.</p></li>
</ul>
<p>The kmer-information is also indicative. If the <span class="bold">FamilyKmers</span> and <span class="bold">KmerCoverage</span> values are low and
the <span class="bold">KmerDupRate</span> value is high, the assigment of the family is only based on a small number of kmers within the reads</p>
<dl class="footnote brackets">
<dt class="label" id="id3"><span class="brackets"><a class="fn-backref" href="#id2">1</a></span></dt>
<dd><p><a class="reference external" href="http://ftp.eva.mpg.de/neandertal/Hohlenstein-Stadel/README">http://ftp.eva.mpg.de/neandertal/Hohlenstein-Stadel/README</a></p>
</dd>
</dl>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="quickstart.html" class="btn btn-neutral float-left" title="Quickstart" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="usage.html" class="btn btn-neutral float-right" title="Usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Merlin Szymanski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>