

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Usage &mdash; quicksand v1.3 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Configuration" href="configuration.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> quicksand
          

          
          </a>

          
            
            
              <div class="version">
                v1.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="description.html">Description</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#flags">Flags</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#required">Required</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#nextflow">Nextflow</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#input">Input</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-pipeline">Run the pipeline</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#run-from-local-repository">Run from local repository</a></li>
<li class="toctree-l3"><a class="reference internal" href="#run-from-remote-repository">Run from remote repository</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#structured-by-taxon">Structured by Taxon</a></li>
<li class="toctree-l3"><a class="reference internal" href="#structured-by-readgroup">Structured by Readgroup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#files-explained">Files explained</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="contribute.html">Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">quicksand</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="usage">
<span id="usage-page"></span><h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="flags">
<h2>Flags<a class="headerlink" href="#flags" title="Permalink to this headline">¶</a></h2>
<div class="section" id="required">
<h3>Required<a class="headerlink" href="#required" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Database-Flags:</strong></div>
<div class="line">(see the <a class="reference internal" href="installation.html#setup"><span class="std std-ref">set-up section</span></a> to create these datastructures)</div>
</div>
<dl class="option-list">
<dt><kbd><span class="option">--db <var>DIR</var></span></kbd></dt>
<dd><p>The preindexed <code class="docutils literal notranslate"><span class="pre">Kraken</span></code> or <code class="docutils literal notranslate"><span class="pre">KrakenUniq</span></code> database used for the classification of reads</p>
</dd>
<dt><kbd><span class="option">--genome <var>DIR</var></span></kbd></dt>
<dd><p>The directory containing the reference genomes in the format <code class="file docutils literal notranslate"><span class="pre">DIR/$family/$species.fasta</span></code></p>
</dd>
<dt><kbd><span class="option">--bedfiles <var>DIR</var></span></kbd></dt>
<dd><p>The directory containing the dustmasked bedfiles in the format <code class="file docutils literal notranslate"><span class="pre">DIR/$species_masked.bed</span></code></p>
</dd>
</dl>
<div class="line-block">
<div class="line"><strong>Input-Flags:</strong></div>
<div class="line">(See <a class="reference internal" href="#input"><span class="std std-ref">Input</span></a>)</div>
</div>
<dl class="option-list">
<dt><kbd><span class="option">--split <var>DIR</var></span></kbd></dt>
<dd><p>The directory containing the input files in the format <code class="file docutils literal notranslate"><span class="pre">DIR/*.{bam|fastq}</span></code></p>
</dd>
<dt><kbd><span class="option">--bam <var>BAM-FILE</var></span></kbd></dt>
<dd><p>A still multiplexed bam-file, containing all the reads</p>
</dd>
<dt><kbd><span class="option">--rg <var>TSV-FILE</var></span></kbd></dt>
<dd><p>The expected readgroups and primer combinations to demultiplex the BAM_FILE by. Fileformat: <code class="code docutils literal notranslate"><span class="pre">RG\\tP7\\p5</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For the the internal use (MPI EVA), there is a demultiplexer available in the pipeline. It is thus possible to run quicksand with <code class="code docutils literal notranslate"><span class="pre">-bam</span> <span class="pre">BAM-FILE</span></code> and <code class="code docutils literal notranslate"><span class="pre">--rg</span> <span class="pre">TSV-FILE</span></code> as alternative to the <code class="code docutils literal notranslate"><span class="pre">--split</span> <span class="pre">DIR</span></code> parameter.</p>
</div>
</div>
<div class="section" id="optional">
<h3>Optional<a class="headerlink" href="#optional" title="Permalink to this headline">¶</a></h3>
<p><strong>Pipeline parameters</strong></p>
<dl class="option-list">
<dt><kbd><span class="option">--analyze</span></kbd></dt>
<dd><p>Given this flag, quicksand will check all alignments in <code class="file docutils literal notranslate"><span class="pre">bedfilteredReads.bam</span></code> for C to T substitutions compared to the reference genome. The file <code class="file docutils literal notranslate"><span class="pre">$RG_deamination_stats.tsv</span></code> is created for each readgroup. See the <a class="reference internal" href="#files"><span class="std std-ref">Files explained</span></a> section. This step takes some time.</p>
</dd>
<dt><kbd><span class="option">--report</span></kbd></dt>
<dd><p>Provided with this flag, quicksand will output a ‘final’ summary called <code class="file docutils literal notranslate"><span class="pre">final_report.tsv</span></code> that summarizes all the stats scattered across the <code class="file docutils literal notranslate"><span class="pre">stats</span></code> directory.</p>
</dd>
<dt><kbd><span class="option">--byrg</span></kbd></dt>
<dd><p>Change the order of output-files in the <code class="file docutils literal notranslate"><span class="pre">out</span></code> directory from a taxonomy-based structure to a readgroup-based structure. See the <a class="reference internal" href="#output"><span class="std std-ref">Output</span></a> section for a comparison</p>
</dd>
<dt><kbd><span class="option">--specmap <var>TSV-FILE</var></span></kbd></dt>
<dd><p>With the TSV_FILE containing families and species in the format <code class="docutils literal notranslate"><span class="pre">Family\tSpecies_name,Species_name</span></code> in each line, ignore the “best taxon” assigned by the kraken classifier as reference for mapping for given families. Instead map the <code class="file docutils literal notranslate"><span class="pre">extractedReads</span></code> of the $family to the species specified in the TSV-FILE. The <code class="docutils literal notranslate"><span class="pre">Species_name</span></code> must correspond to a filename (without <code class="code docutils literal notranslate"><span class="pre">.fasta</span></code>) in the <code class="file docutils literal notranslate"><span class="pre">genomes/$family/</span></code> directory</p>
</dd>
</dl>
<p>–taxlvl [f,o]    bin reads assigned by krakenuniq by the specified taxon level (family or order level). Extracted reads ar later mapped against the genome of each families best fitting representative species. For the order-level extract, reads are thus mapped several times to different genomes. For family-level extracts, reads are mapped only once. [default: family level]
–capture STRING    Given a STRING <code class="docutils literal notranslate"><span class="pre">Family,Family,...</span></code>, apply reduced filters to the given families. For provided families, the bedfiltering is ommited</p>
<p><strong>Process parameters</strong></p>
<dl class="option-list">
<dt><kbd><span class="option">--cutoff <var>N</var></span></kbd></dt>
<dd><p>Length cutoff after BWA-mapping (default: 35). Remove all reads below a length of N from the alignemnt.</p>
</dd>
<dt><kbd><span class="option">--quality <var>N</var></span></kbd></dt>
<dd><p>Mapping quality filter after BWA-mapping (default:25). Remove all reads with a mapping-quality of &lt;N from the alignment.</p>
</dd>
<dt><kbd><span class="option">--min_kmers <var>N</var></span></kbd></dt>
<dd><p>Minimum required unique kmers assigned by KrakenUniq (default: 129). To reduce false-positives: Remove families from the assignments if the total number of assigned unique kmers to taxa of that family by KrakenUniq is below N</p>
</dd>
<dt><kbd><span class="option">--min_reads <var>N</var></span></kbd></dt>
<dd><p>Minimum required assigned reads by KrakenUniq (default: 3). To reduce false-positives: Remove families from the assignments if the total number of assigned reads to taxa of that family by KrakenUniq is below N</p>
</dd>
<dt><kbd><span class="option">--keeppaired</span></kbd></dt>
<dd><p>If <code class="file docutils literal notranslate"><span class="pre">.bam</span></code> files are provided as input, by default, unmerged reads are filtered out. This flag prevents that.</p>
</dd>
<dt><kbd><span class="option">--filterunmapped</span></kbd></dt>
<dd><p>If premapped <code class="file docutils literal notranslate"><span class="pre">.bam</span></code> files are provided as input, use this flag to filter out unmapped reads from the bam-files</p>
</dd>
<dt><kbd><span class="option">--level <var>N</var></span></kbd></dt>
<dd><p>Set BGZF compression level (default: 6)</p>
</dd>
<dt><kbd><span class="option">--krakenthreads <var>N</var></span></kbd></dt>
<dd><p>Number of threads per Kraken process (default: 4)</p>
</dd>
</dl>
</div>
<div class="section" id="nextflow">
<h3>Nextflow<a class="headerlink" href="#nextflow" title="Permalink to this headline">¶</a></h3>
<p>A selection of built-in Nextflow flags that may be of use (Be aware: only one dash - with these flags):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">profile</span>  <span class="o">&lt;</span><span class="n">profile</span><span class="o">&gt;</span>  <span class="n">pick</span> <span class="n">a</span> <span class="n">profile</span> <span class="k">for</span> <span class="n">executing</span> <span class="n">the</span> <span class="n">pipeline</span> <span class="p">(</span><span class="n">see</span> <span class="n">below</span><span class="p">)</span>
<span class="o">-</span><span class="n">resume</span>              <span class="n">Resume</span> <span class="n">processing</span><span class="p">;</span> <span class="n">do</span> <span class="ow">not</span> <span class="n">re</span><span class="o">-</span><span class="n">run</span> <span class="n">completed</span> <span class="n">processes</span>
<span class="o">-</span><span class="n">N</span>        <span class="o">&lt;</span><span class="n">email</span><span class="o">&gt;</span>    <span class="n">send</span> <span class="n">notification</span> <span class="n">to</span> <span class="n">email</span> <span class="n">upon</span> <span class="n">finishing</span>
<span class="o">-</span><span class="n">c</span>        <span class="o">&lt;</span><span class="n">file</span><span class="o">&gt;</span>     <span class="n">path</span> <span class="n">to</span> <span class="n">additional</span> <span class="n">nextflow</span><span class="o">.</span><span class="n">config</span> <span class="n">file</span> <span class="k">for</span> <span class="n">local</span> <span class="n">parameters</span>
<span class="o">-</span><span class="n">w</span>        <span class="o">&lt;</span><span class="n">path</span><span class="o">&gt;</span>     <span class="n">specify</span> <span class="n">the</span> <span class="s2">&quot;work&quot;</span> <span class="n">directory</span> <span class="k">for</span> <span class="n">the</span> <span class="n">nextflow</span> <span class="n">intermediate</span> <span class="n">files</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="input">
<span id="id1"></span><h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<p>The pipeline uses as input <code class="file docutils literal notranslate"><span class="pre">.fastq</span></code> or <code class="file docutils literal notranslate"><span class="pre">.bam</span></code> files that contain demultiplexed, merged, and adapter trimmed reads.
Use the <code class="code docutils literal notranslate"><span class="pre">--split</span></code> flag to point to the directory that contains these files. The pipeline refers to the name of the files
as readgroups. The reads within the files are assigned, processed and structured by readgroups.
<code class="file docutils literal notranslate"><span class="pre">.bam</span></code> and <code class="file docutils literal notranslate"><span class="pre">.fastq</span></code> files can be mixed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">splitdir</span><span class="o">/</span>
    <span class="n">readgroup1</span><span class="o">.</span><span class="n">fastq</span>
    <span class="n">readgroup2</span><span class="o">.</span><span class="n">fastq</span>
    <span class="n">readgroup3</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Quicksand will process all the <code class="file docutils literal notranslate"><span class="pre">.fastq</span></code> and <code class="file docutils literal notranslate"><span class="pre">.bam</span></code>-files within the <code class="code docutils literal notranslate"><span class="pre">--split</span></code> directory and ignore all remaining files.
Be sure to name/rename your files accordingly (e.g rename <code class="file docutils literal notranslate"><span class="pre">.fq</span></code> to <code class="file docutils literal notranslate"><span class="pre">.fastq</span></code> files)</p>
</div>
</div>
<div class="section" id="run-the-pipeline">
<h2>Run the pipeline<a class="headerlink" href="#run-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>As outlined below, there are several ways to run the pipeline.
Please see to the <a class="reference internal" href="installation.html#install-page"><span class="std std-ref">installation</span></a> to set up the required underying databases and
the <a class="reference internal" href="#input"><span class="std std-ref">input section</span></a> to know about the required input.</p>
<div class="admonition attention">
<p class="admonition-title">Attention</p>
<p>Before starting the pipeline make sure that you are in separate folder. The pipeline writes its output into the current working directory!</p>
</div>
<div class="section" id="run-from-local-repository">
<h3>Run from local repository<a class="headerlink" href="#run-from-local-repository" title="Permalink to this headline">¶</a></h3>
<p>With everything set up quicksand can be executed by pointing nextflow to the <code class="file docutils literal notranslate"><span class="pre">main.nf</span></code> file in the cloned repository:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">runDir</span> <span class="o">&amp;&amp;</span> <span class="n">cd</span> <span class="n">runDir</span>

<span class="n">nextflow</span> <span class="n">run</span> <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">quicksand</span><span class="o">/</span><span class="n">main</span><span class="o">.</span><span class="n">nf</span> \
    <span class="o">--</span><span class="n">split</span>     <span class="n">PATH</span><span class="o">/</span><span class="n">TO</span><span class="o">/</span><span class="n">INPUT</span><span class="o">/</span><span class="n">DATA</span> \
    <span class="o">--</span><span class="n">genome</span>    <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">genomes</span> \
    <span class="o">--</span><span class="n">db</span>        <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">kraken</span><span class="o">/</span><span class="n">Mito_db_kmer22</span> \
    <span class="o">--</span><span class="n">bedfiles</span>  <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">masked</span> \
    <span class="o">--</span><span class="n">report</span>
    <span class="o">--</span><span class="n">analyze</span>
</pre></div>
</div>
</div>
<div class="section" id="run-from-remote-repository">
<h3>Run from remote repository<a class="headerlink" href="#run-from-remote-repository" title="Permalink to this headline">¶</a></h3>
<p>Instead of cloning the repository, it is also possible to run the pipeline directly from github:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">runDir</span> <span class="o">&amp;&amp;</span> <span class="n">cd</span> <span class="n">runDir</span>

<span class="n">nextflow</span> <span class="n">run</span> <span class="n">mpieva</span><span class="o">/</span><span class="n">quicksand</span> \
    <span class="o">--</span><span class="n">split</span>     <span class="n">PATH</span><span class="o">/</span><span class="n">TO</span><span class="o">/</span><span class="n">INPUT</span><span class="o">/</span><span class="n">DATA</span> \
    <span class="o">--</span><span class="n">genome</span>    <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">genomes</span> \
    <span class="o">--</span><span class="n">db</span>        <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">kraken</span><span class="o">/</span><span class="n">Mito_db_kmer22</span> \
    <span class="o">--</span><span class="n">bedfiles</span>  <span class="o">~/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">masked</span> \
    <span class="o">--</span><span class="n">report</span>
    <span class="o">--</span><span class="n">analyze</span>
</pre></div>
</div>
<p>It is possible to specify a version (or branch) by providing the <code class="code docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;branch/tag&gt;</span></code> flag. By default the <code class="code docutils literal notranslate"><span class="pre">master</span></code> branch is pulled
and stored locally in the hidden <code class="file docutils literal notranslate"><span class="pre">~/.nextflow</span></code> directory. If the remote pipeline is updated, make sure to pull the updates by running:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">pull</span> <span class="n">mpieva</span><span class="o">/</span><span class="n">quicksand</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span> <span class="n">v1</span><span class="mf">.3</span><span class="p">]</span>
</pre></div>
</div>
<p>To reduce the number of flags manually handed over to the pipeline, see the <a class="reference internal" href="configuration.html#configuration-page"><span class="std std-ref">configuration section</span></a> on how to set up an individual configuration of the pipeline.</p>
</div>
</div>
<div class="section" id="output">
<span id="id2"></span><h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>For each readgroup, the pipeline outputs the raw reads in <code class="file docutils literal notranslate"><span class="pre">.bam</span></code>-format at each stage of the pipeline.
Additionally the number of reads assigned to one taxon-level (Family/Order), reads extracted, reads mapped, reads deduplicated, and bedfiltered as well as the estimated DNA-damage
are reported for a quick overview in one big summary-file.</p>
<p>Two ways exist to structure the output. Based on the <code class="code docutils literal notranslate"><span class="pre">--byrg</span></code> flag either by readgroup or by family.
without the <code class="code docutils literal notranslate"><span class="pre">--byrg</span></code> flag, the output is structured by Taxon</p>
<div class="section" id="structured-by-taxon">
<h3>Structured by Taxon<a class="headerlink" href="#structured-by-taxon" title="Permalink to this headline">¶</a></h3>
<p>This overview corresponds to a run with the <code class="code docutils literal notranslate"><span class="pre">--analyze</span></code> and the <code class="code docutils literal notranslate"><span class="pre">--report</span></code> flags provided.
Several directories and files should appear after the run. The ‘taxon’ corresponds to  either the family or the order level:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>RunDir
├── out
│    └── {taxon}
│         ├── {readgroup}_extractedReads-{family}.bam
│         ├── aligned
│         │    ├── {readgroup}.{family}.{species}.bam
│         │    └── {readgroup}.{family}.{species}_deduped.bam
│         └── bed
│              └── {readgroup}.{family}.{species}_deduped_bedfiltered.bam
├── kraken
│    ├── {readgroup}.report
│    └── {readgroup}.translate
├── stats
│    ├── splitcounts.tsv
│    ├── {readgroup}_extracted.tsv
│    ├── {readgroup}_mapped.tsv
│    ├── {readgroup}_mapped_coverage.tsv
│    ├── {readgroup}_unique_mapped.tsv
│    ├── {readgroup}_bedfiltered.tsv
│    └── {readgroup}_deamination_stats
├── reports
│    ├── report.html
│    ├── timeline.html
│    └── trace.tsv
├── work
│    └── ...
└── final_report.tsv
</pre></div>
</div>
</div>
<div class="section" id="structured-by-readgroup">
<h3>Structured by Readgroup<a class="headerlink" href="#structured-by-readgroup" title="Permalink to this headline">¶</a></h3>
<p>This overview of the <code class="file docutils literal notranslate"><span class="pre">out/</span></code> dir corresponds to a <code class="code docutils literal notranslate"><span class="pre">--byrg</span></code> run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>RunDir
├── out
│    └── {readgroup}
│         ├── {readgroup}_extractedReads-{taxon}.bam
│         ├── aligned
│         │    ├── {readgroup}.{family}.{species}.bam
│         │    └── {readgroup}.{family}.{species}_deduped.bam
│         └── bed
│              └── {readgroup}.{family}.{species}_deduped_bedfiltered.bam
...
</pre></div>
</div>
</div>
<div class="section" id="files-explained">
<span id="files"></span><h3>Files explained<a class="headerlink" href="#files-explained" title="Permalink to this headline">¶</a></h3>
<p>The content of the files is explained here:</p>
<p><strong>out/</strong></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">extractedReads.bam</span></code></p></td>
<td><p>Contains all DNA sequences of one family (or order) from one readgroup assigned by KrakenUniq to that taxon.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="file docutils literal notranslate"><span class="pre">alignedReads.bam</span></code></p></td>
<td><p>For a given family and species, all extractedReads (see above) mapped/aligned to the genome of the assigned species</p></td>
</tr>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">aligned_dedupedReads.bam</span></code></p></td>
<td><p>The aligned reads, but depleted of PCR duplicates</p></td>
</tr>
<tr class="row-odd"><td><p><code class="file docutils literal notranslate"><span class="pre">bedfilteredReads</span></code></p></td>
<td><p>The aligned_deduped reads but additionally depleted of reads overlapping low-complexity regions</p></td>
</tr>
</tbody>
</table>
<p><strong>kraken/</strong></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">kraken.report</span></code></p></td>
<td><p>The raw KrakenUniq report</p></td>
</tr>
<tr class="row-odd"><td><p><code class="file docutils literal notranslate"><span class="pre">kraken.translate</span></code></p></td>
<td><p>The human readable kraken report in mpa-format</p></td>
</tr>
</tbody>
</table>
<p><strong>stats/</strong></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">splitcounts.tsv</span></code></p></td>
<td><p>Contains the number of reads per readgroup</p></td>
</tr>
<tr class="row-odd"><td><p><code class="file docutils literal notranslate"><span class="pre">extracted.tsv</span></code></p></td>
<td><p>For each readgroup show the number of reads per extracted family</p></td>
</tr>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">mapped.tsv</span></code></p></td>
<td><p>Show the number of reads mapped against the reference genome</p></td>
</tr>
<tr class="row-odd"><td><p><code class="file docutils literal notranslate"><span class="pre">mapped_coverage.tsv</span></code></p></td>
<td><p>Contains the number of covered basepairs for each mapping</p></td>
</tr>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">unique_mapped.tsv</span></code></p></td>
<td><p>Contains the number of deduplicated reads mapped against the reference genome</p></td>
</tr>
<tr class="row-odd"><td><p><code class="file docutils literal notranslate"><span class="pre">bedfiltered.tsv</span></code></p></td>
<td><p>The number of reads remaining in the bam-file after bedfiltering</p></td>
</tr>
<tr class="row-even"><td><p><code class="file docutils literal notranslate"><span class="pre">deamination_stats</span></code></p></td>
<td><p>Show an estimation of the ‘ancientness’ of families based on deamination frequencies of recovered reads</p></td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line">The <code class="file docutils literal notranslate"><span class="pre">final_report.tsv</span></code> is a summary of all the files in the <code class="file docutils literal notranslate"><span class="pre">stats</span></code> dir.</div>
<div class="line">The <code class="file docutils literal notranslate"><span class="pre">report</span></code> directory contains information about the run</div>
<div class="line">the <code class="file docutils literal notranslate"><span class="pre">work</span></code> directory can be deleted after the run - it contains nextflow specific intermediate files</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="configuration.html" class="btn btn-neutral float-right" title="Configuration" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, MPI EVA.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>